content/blog/2019-10-23-loop-unswitching/index.md

```diff
@@ -0,0 +1,252 @@
+    
+In one body block, we include the contents of the ```if``` block, and in the other body block, we include the contents of the ```else``` block.  Finally, we add instructions that are dominated by each of those blocks to be part of the body block.  Combining all of this, as well as performing surgery on the branch instruction names completes the construction of the ```if loop body``` and ```else loop body``` blocks.
+
+Next, we create the additional blocks, namely the ```jmp``` and ```bypass``` blocks that lead to the end of the program.  Creating these blocks requires surgery on the original instructions, as well as rename hashing, since it is not always guaranteed that the last instruction in a block is a ```jump``` instruction.  For instance, both of the following are valid Bril programs:
+```
+blockone:
+    x: int = const 0;
+    jmp blocktwo;
+blocktwo:
+    print x;
+```
+
+and
+```
+blockone:
+    x: int = const 0;
+blocktwo:
+    print x;
+```
+so we need to guarantee that these two blocks are placed next to each other when we output the new optimized program.  
+
+Finally, we connect the bypass blocks to the end of the program, which we leave intact.   
+
+The last step is to overwrite the original mapping from block name to contents with our new block dictionary and output the resulting Bril program.
+
+## Implementation Difficulties
+
+There were quite a few difficulties that arose during implementation due to the nature of the Bril programming language.
+
+As alluded to before, since Bril does not require jumps at the end of blocks, we need to be careful about reordering blocks in the final Bril program.  Furthermore, this optimization is designed to run after other optimizations, so we would ideally like to keep the majority of the program untouched to prevent overwriting.  This involves a fair amount of bookkeeping especially when these situations occur within the loop body.  
+
+Traditionally, loop unswitching operates on SSA form, and due to time restriction, I was not able to write an SSA translater.  Thus this optimization assumes a prior SSA run.  One example of why we might want SSA form is dead code elimination.  It is possible that after dead code elimination, a branch within a loop becomes independent of the loop body.
+
+Another shortcoming of this optimization is that it currently only operates on natural loops, and in particular, does not operate on nested loops.  That is, programs in the form:
+```
+b = True
+for i in range(10):
+    for j in range(10):
+        if b then:
+            do something()
+        else:
+            do something else()
+```
+will not be completely unswitched since the conditional statement lies inside of a nested for loop.  Instead, the conditional will be unswitched outside of the first for loop, which only midly decreases the numbeer of branches.  Implementing this is not too difficult to do, though it requires  additional abstractions for nested for loops.
```

Comment:
```suggestion
will not be completely unswitched since the conditional statement lies inside of a nested for loop.  Instead, the conditional will be unswitched outside of the first for loop, which only midly decreases the number of branches.  Implementing this is not too difficult to do, though it requires  additional abstractions for nested for loops.
```
Please run spell check.